### ISSUE TITLE
FP16 brings inconsistent BERT results every time

### ISSUE THREAD

**Issue by oujiafan** [NVIDIA/TensorRT#568]
## Description

I used the BERT demo code.  When I enabled fp16, it returned different scores on different runs. It seems not to impact the model accuracy but brings inconsistency/stability issue for our service.
Is it a known issue for TensorRT? Is there any option to make it stable when it optimizes model at runtime?

I also tried to save the engine and load the engine every time. It can give us consistent result. However, the same engine loaded on Linux and Windows can return different scores.

## Environment

**TensorRT Version**: 5/6/7
**GPU Type**: P100/V100
**Nvidia Driver Version**: Whatever from 10.0/10.1/10.2
**CUDA Version**: 10.0/10.1/10.2
**CUDNN Version**: 7.6.5
**Operating System + Version**: Windows/Ubuntu
**Python Version (if applicable)**: 
**TensorFlow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if container which image + tag)**:

**Comment by rmccorm4** [comment]
> I also tried to save the engine and load the engine every time. It can give us consistent result. 

Engine files are generally unique to OS, Platform, GPU type, GPU memory, etc.. Once the engine has been created, the results should be consistent on that machine that the engine was created on. However, the engine is not portable across these parameters, including across Windows/Linux. You'll want to create a separate engine file on each machine, optimized for the machine it was created on.

> Is there any option to make it stable when it optimizes model at runtime?

Currently in TensorRT 7.0, CUDA kernels (tactics) for each op/layer are timed at build time to see which kernels are fastest on the machine it is being built on. It is possible that some tactics are very close to each other, and you may get different tactics selected each build.

In TensorRT 7.1 (currently in Early Access for Jetson only), I believe the new Algorithm Selection API will allow you to choose your tactics to build the same engine deterministically each time: https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#algorithm-select. This should be available once TensorRT 7.1 is fully released on platforms other than Jetson.

**Comment by oujiafan** [comment]
Thank you @rmccorm4 !
Good to hear we have some flexibility in TRT 7.1. Is it possible to introduce some simple stability option to existing version or 7.1? E.g., when several tactics are very close we can just choose first one by a stable order (say, by name).

"you may get different tactics selected each build" -- I suspect either we don't have stable measurement on the tactics testing (we should run many time to get stable result), or they're simply equal. When they're equal we should do a stable selection, not random selection. How do you think?

We need consistent score across different OS. Is there any other option?

**Comment by rmccorm4** [comment]
I'm not sure on all of the details at the moment as I haven't used the new API yet. I would guess you can build an engine in a loop several times with VERBOSE logging, take note of frequency of which tactics are selected, and use the IAlgorithmSelector to choose the most frequent tactic each time (on both OS's if you want the same tactics on both).

**Comment by oujiafan** [comment]
I was questioning the tactics selection on existing versions. It should be stable on the same machine. But it's not the case now.

**Comment by oujiafan** [comment]
Do we know why CUDA 10.2+TRT 7.0 on Ubuntu 18 can give consistent score? Is there any bug in Windows version?

**Comment by asmasaad** [comment]
I have the same issue, when I convert the model from onnx to trt, I got mismatch result, Also I got different result when engine deploy on jetson or on ubuntu with GPU

**Comment by oujiafan** [comment]
A solution could be: save as engine file for different platform. So machines in same platform can get same result.

**Comment by Semen52** [comment]
> A solution could be: save as engine file for different platform. So machines in same platform can get same result.

If you mean "the same GPU card", yes you're right.
